### PR TITLE
Teams to tournament

### PR DESCRIPTION
**Pull request by IgorLeonenko**
This change is [Reviewable](https://reviewable.io/reviews/igorleonenko/matches/5)

### CODE REVIEW

**Comment by mskv** [comment]
Reviewed 8 of 8 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

_[app/controllers/tournaments_controller.rb, line 9 at r1](https://reviewable.io:443/reviews/igorleonenko/matches/5#-KSkm5sag72QfCU1xqBg:-KSkm5sbkPi8198Ss_oN:b-845p58) ([raw file](https://github.com/igorleonenko/matches/blob/13aa61eaebbf612956e7d25bd03d438a163fab24/app/controllers/tournaments_controller.rb#L9)):_

> ``` Ruby
> 
>   def show
>     @teams = Team.all
> ```

I don't think it is used in the view. In the view we were referencing teams using `@tournament.teams`.

---

_[app/controllers/tournaments_controller.rb, line 21 at r1](https://reviewable.io:443/reviews/igorleonenko/matches/5#-KSklfOK9qO0fmkKj18Y:-KSklfOK9qO0fmkKj18Z:b-b7ryw) ([raw file](https://github.com/igorleonenko/matches/blob/13aa61eaebbf612956e7d25bd03d438a163fab24/app/controllers/tournaments_controller.rb#L21)):_

> ``` Ruby
>   end
> 
>   def team
> ```

Please check out
http://jeromedalbert.com/how-dhh-organizes-his-rails-controllers/

It's not mandatory - but it's a pattern. The idea is to stick to the CRUD. 

So if we're adding users to a team, we could either use
`TeamsController#update`
or create a new controller
`UserTeamMembershipsController`
and define a `create` action there.

The idea is that Model != Controller. There could be abstract controllers. The only thing to remember - stick to the CRUD.

So to sum up:
1. `def team` should probably be `TeamsController#new`
2. `def join_team` should probably be either `TeamsController#update` or `UserTeamMembershipsController#create`

Again - it's a pattern. Maybe you don't want to use this pattern - that's OK.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/igorleonenko/matches/5)_

**Comment by mskv** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

_[app/controllers/tournaments_controller.rb, line 45 at r1](https://reviewable.io:443/reviews/igorleonenko/matches/5#-KSkojsEup0VxqVwIg2C:-KSkojsEup0VxqVwIg2D:b-jag387) ([raw file](https://github.com/igorleonenko/matches/blob/13aa61eaebbf612956e7d25bd03d438a163fab24/app/controllers/tournaments_controller.rb#L45)):_

> ``` Ruby
>   end
> 
>   def add_user_to_team
> ```

I think this empty method is used in the view. Maybe `join_team` should have been used here instead?

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/igorleonenko/matches/5)_

**Comment by IgorLeonenko** [comment]
_[app/controllers/tournaments_controller.rb, line 21 at r1](https://reviewable.io:443/reviews/igorleonenko/matches/5#-KSklfOK9qO0fmkKj18Y:-KSoospmZsqCdrIph-i5:bt8b5p6) ([raw file](https://github.com/igorleonenko/matches/blob/13aa61eaebbf612956e7d25bd03d438a163fab24/app/controllers/tournaments_controller.rb#L21)):_
<details><summary>

<i>Previously, mskv (Michał Podwórny) wrote…</i></summary>



> Please check out
> http://jeromedalbert.com/how-dhh-organizes-his-rails-controllers/
> 
> It's not mandatory - but it's a pattern. The idea is to stick to the CRUD. 
> 
> So if we're adding users to a team, we could either use
> `TeamsController#update`
> or create a new controller
> `UserTeamMembershipsController`
> and define a `create` action there.
> 
> The idea is that Model != Controller. There could be abstract controllers. The only thing to remember - stick to the CRUD.
> 
> So to sum up:
> 1. `def team` should probably be `TeamsController#new`
> 2. `def join_team` should probably be either `TeamsController#update` or `UserTeamMembershipsController#create`
> 
> Again - it's a pattern. Maybe you don't want to use this pattern - that's OK.
> </details>
> yep! Now I see, that it's be better :)

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/igorleonenko/matches/5)_

**Comment by mkarbowiak** [comment]
Review status: 0 of 23 files reviewed at latest revision, 14 unresolved discussions.

---

_[app/controllers/teams_controller.rb, line 4 at r2](https://reviewable.io:443/reviews/igorleonenko/matches/5#-KT8ekfeyP9kUQA35bmW:-KT8ekfeyP9kUQA35bmX:b2k7b2p) ([raw file](https://github.com/igorleonenko/matches/blob/15b185f040dd953810b7bdbcfcc6a3568086ff65/app/controllers/teams_controller.rb#L4)):_

> ``` Ruby
>   before_action :set_users, only: [:new, :create]
>   before_action :set_tournament
>   before_action :set_team, only: :update
> ```

Why use those `set_` methods instead of, for example, lazy loading? Is there a specific reason?

---

_[app/controllers/teams_controller.rb, line 13 at r2](https://reviewable.io:443/reviews/igorleonenko/matches/5#-KT8g8Z4k-XOZ5F1IVhG:-KT8g8Z4k-XOZ5F1IVhH:b-v6teti) ([raw file](https://github.com/igorleonenko/matches/blob/15b185f040dd953810b7bdbcfcc6a3568086ff65/app/controllers/teams_controller.rb#L13)):_

> ``` Ruby
>     @team = Team.new(team_params)
>     users_to_team = params[:team][:user_ids]
>     unless users_to_team.blank?
> ```

Isn't `if something` more readable than `unless`?

---

_[app/controllers/teams_controller.rb, line 16 at r2](https://reviewable.io:443/reviews/igorleonenko/matches/5#-KT8grhreeJNwt3H-Lbg:-KT8grhreeJNwt3H-Lbh:b1usvdy) ([raw file](https://github.com/igorleonenko/matches/blob/15b185f040dd953810b7bdbcfcc6a3568086ff65/app/controllers/teams_controller.rb#L16)):_

> ``` Ruby
>       users_to_team.each do |user|
>         @team.users << User.find(user)
>       end
> ```

You named this variable `user` but I think it's more `user_id`, right? `user` suggest instance of User class, not id. 
You could actually shrink this create method with some decoupling.
For example:
1. Create some private method `assign_users_to_team`
2. In that method check if `params[:team][:user_ids]` is empty or not (present or not) using **guard clause**
3. Iterate over those ids and assign them as you're doing here
4. In `create` method here, just invoke `assign_users_to_team`
5. ???
6. Smaller and more abstract create method, better readability, simple code

You can try and use one-liners where they'll fit 100 columns, for example:
`user_to_teams.each { |user_id| @team.users << User.find(user_id) }`

---

_[app/controllers/teams_controller.rb, line 23 at r2](https://reviewable.io:443/reviews/igorleonenko/matches/5#-KT8tqRRh3e4B8tk0ow4:-KT8tqRRh3e4B8tk0ow5:bvn0qv4) ([raw file](https://github.com/igorleonenko/matches/blob/15b185f040dd953810b7bdbcfcc6a3568086ff65/app/controllers/teams_controller.rb#L23)):_

> ``` Ruby
>         @tournament.add_team!(@team)
>         @team.save!
>         @tournament.add_user!(@team)
> ```

You're invoking `add_team!` method first, then `save!` on `@team`, why in this order? That if saving of team fails? Shouldn't this be handled before?

---

_[app/controllers/teams_controller.rb, line 25 at r2](https://reviewable.io:443/reviews/igorleonenko/matches/5#-KT8wdtXRu82pYW2G9cG:-KT8wdtXRu82pYW2G9cH:bebjg7w) ([raw file](https://github.com/igorleonenko/matches/blob/15b185f040dd953810b7bdbcfcc6a3568086ff65/app/controllers/teams_controller.rb#L25)):_

> ``` Ruby
>         @tournament.add_user!(@team)
>         flash[:notice] = 'Team added'
>       rescue => e
> ```

If you're rescuing an error here...

---

_[app/controllers/teams_controller.rb, line 29 at r2](https://reviewable.io:443/reviews/igorleonenko/matches/5#-KT8wfzx7Wfz014VjZ_4:-KT8wfzx7Wfz014VjZ_5:bgofj1z) ([raw file](https://github.com/igorleonenko/matches/blob/15b185f040dd953810b7bdbcfcc6a3568086ff65/app/controllers/teams_controller.rb#L29)):_

> ``` Ruby
>         flash.now[:alert] = "#{e}"
>       end
>       raise ActiveRecord::Rollback if @error
> ```

..why are you raising another one here?

---

_[app/controllers/teams_controller.rb, line 40 at r2](https://reviewable.io:443/reviews/igorleonenko/matches/5#-KT8w9HGb8BiodCWh1I7:-KT8w9HGb8BiodCWh1I8:b-f5qi76) ([raw file](https://github.com/igorleonenko/matches/blob/15b185f040dd953810b7bdbcfcc6a3568086ff65/app/controllers/teams_controller.rb#L40)):_

> ``` Ruby
> 
>   def update
>     user = User.find(params[:team][:user_ids])
> ```

You're using `user_ids` so shouldn't this variable be named `users`? Or is this just single id? 

---

_[app/controllers/teams_controller.rb, line 50 at r2](https://reviewable.io:443/reviews/igorleonenko/matches/5#-KT8wlE4IGOfi8jzyasR:-KT8wlE4IGOfi8jzyasS:brhumzt) ([raw file](https://github.com/igorleonenko/matches/blob/15b185f040dd953810b7bdbcfcc6a3568086ff65/app/controllers/teams_controller.rb#L50)):_

> ``` Ruby
>         flash[:alert] = "#{e}"
>       end
>       raise ActiveRecord::Rollback if @error
> ```

Same here.

---

_[app/controllers/teams_controller.rb, line 53 at r2](https://reviewable.io:443/reviews/igorleonenko/matches/5#-KT9-4n-JRFU98X3ycD_:-KT9-4n-JRFU98X3ycDa:bjlks1t) ([raw file](https://github.com/igorleonenko/matches/blob/15b185f040dd953810b7bdbcfcc6a3568086ff65/app/controllers/teams_controller.rb#L53)):_

> ``` Ruby
>     end
>     redirect_to @tournament
>   end
> ```

And what if we want to update the name of the team? Or are we not allowed to?

---

_[app/models/tournament.rb, line 25 at r2](https://reviewable.io:443/reviews/igorleonenko/matches/5#-KT8xDhbdxVh69kTGQuf:-KT8xDhbdxVh69kTGQug:bn2m9u4) ([raw file](https://github.com/igorleonenko/matches/blob/d2712116aa05e04cfa227538baa67e75ff7817fc/app/models/tournament.rb#L25)):_

> ``` Ruby
>   validates :style, inclusion: { in: STYLE }
>   validates :state, inclusion: { in: STATE }
>   validate  :cant_be_more_teams_than_teams_quantity
> ```

Why did you resign from custom validation? It's a lot better than raising own errors like you're doing down below. 

---

_[app/models/tournament.rb, line 37 at r2](https://reviewable.io:443/reviews/igorleonenko/matches/5#-KT8xX77fuZvnB3FE91Z:-KT8xX77fuZvnB3FE91_:b9n3dvi) ([raw file](https://github.com/igorleonenko/matches/blob/15b185f040dd953810b7bdbcfcc6a3568086ff65/app/models/tournament.rb#L37)):_

> ``` Ruby
>   def add_team!(team)
>     if teams.size >= teams_quantity
>       raise 'Validation failed: Can\'t be more teams than teams quantity'
> ```

Should be custom validation, then.
After that, you'll have all errors in `resource.errors` method.

---

_[app/models/tournament.rb, line 45 at r2](https://reviewable.io:443/reviews/igorleonenko/matches/5#-KT9000vVKUl1ctSk7xA:-KT9000vVKUl1ctSk7xB:b9lawpq) ([raw file](https://github.com/igorleonenko/matches/blob/15b185f040dd953810b7bdbcfcc6a3568086ff65/app/models/tournament.rb#L45)):_

> ``` Ruby
>   def add_user!(team)
>     if !players_in_team.nil? && team.users.size > players_in_team
>       raise 'Validation failed: Can\'t be more players than players in team'
> ```

And here. Actually, can you explain what this method is doing?

---

_[spec/controllers/teams_controller_spec.rb, line 14 at r2](https://reviewable.io:443/reviews/igorleonenko/matches/5#-KT90NqqojoglaE4Fk5D:-KT90NqqojoglaE4Fk5E:bla15mz) ([raw file](https://github.com/igorleonenko/matches/blob/15b185f040dd953810b7bdbcfcc6a3568086ff65/spec/controllers/teams_controller_spec.rb#L14)):_

> ``` Ruby
>       context 'with valid attributes' do
>         subject { post :create, params: { tournament_id: tournament.id,
>                        team: { name: 'good team', user_ids: [user.id] } } }
> ```

Always use `do / end` with multiline blocks.

---

_[spec/controllers/teams_controller_spec.rb, line 61 at r2](https://reviewable.io:443/reviews/igorleonenko/matches/5#-KT91l7chZ5NbQIpbvKB:-KT91l7dcyXwPdwEwTtx:bvgojbf) ([raw file](https://github.com/igorleonenko/matches/blob/15b185f040dd953810b7bdbcfcc6a3568086ff65/spec/controllers/teams_controller_spec.rb#L61)):_

> ``` Ruby
>               tournament.users << u
>             end
>             tournament.save
> ```

Instead of performing actions like these, it's better to create a new `let`s with proper data.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/igorleonenko/matches/5)_

**Comment by IgorLeonenko** [comment]
Review status: 0 of 23 files reviewed at latest revision, 14 unresolved discussions.

---

_[app/controllers/teams_controller.rb, line 4 at r2](https://reviewable.io:443/reviews/igorleonenko/matches/5#-KT8ekfeyP9kUQA35bmW:-KT98d_OfrmvcXM-sCol:b-s9mw) ([raw file](https://github.com/igorleonenko/matches/blob/15b185f040dd953810b7bdbcfcc6a3568086ff65/app/controllers/teams_controller.rb#L4)):_
<details><summary>

<i>Previously, mkarbowiak (Mateusz Karbowiak) wrote…</i></summary>



> Why use those `set_` methods instead of, for example, lazy loading? Is there a specific reason?
> </details>
> what can I use instead for example:
> def set_team
>   @team = Team.find(params[:id]
> end
> ?

---

_[app/controllers/teams_controller.rb, line 13 at r2](https://reviewable.io:443/reviews/igorleonenko/matches/5#-KT8g8Z4k-XOZ5F1IVhG:-KT9Bcu1isGTZYe9ZvLq:bjmo7wr) ([raw file](https://github.com/igorleonenko/matches/blob/15b185f040dd953810b7bdbcfcc6a3568086ff65/app/controllers/teams_controller.rb#L13)):_
<details><summary>

<i>Previously, mkarbowiak (Mateusz Karbowiak) wrote…</i></summary>



> Isn't `if something` more readable than `unless`?
> </details>
> yep! You are right. Must change this things

---

_[app/controllers/teams_controller.rb, line 23 at r2](https://reviewable.io:443/reviews/igorleonenko/matches/5#-KT8tqRRh3e4B8tk0ow4:-KT9D0R0RZ8bpq_sDg58:b-dmag1) ([raw file](https://github.com/igorleonenko/matches/blob/15b185f040dd953810b7bdbcfcc6a3568086ff65/app/controllers/teams_controller.rb#L23)):_
<details><summary>

<i>Previously, mkarbowiak (Mateusz Karbowiak) wrote…</i></summary>



> You're invoking `add_team!` method first, then `save!` on `@team`, why in this order? That if saving of team fails? Shouldn't this be handled before?
> </details>
> It's goes like this because in method add_team! assigns tournament_id to team object. And how I understand if one of methods in transaction failed, the whole transaction will be rollbacked

---

_[app/controllers/teams_controller.rb, line 29 at r2](https://reviewable.io:443/reviews/igorleonenko/matches/5#-KT8wfzx7Wfz014VjZ_4:-KT9GB6MAPMxVtfaUxCo:b-wuu5rz) ([raw file](https://github.com/igorleonenko/matches/blob/15b185f040dd953810b7bdbcfcc6a3568086ff65/app/controllers/teams_controller.rb#L29)):_
<details><summary>

<i>Previously, mkarbowiak (Mateusz Karbowiak) wrote…</i></summary>



> ..why are you raising another one here?
> </details>
> Do it to rollback transaction. I saw this in some tutorial and it's work. But maybe it can be some better solution?

---

_[app/controllers/teams_controller.rb, line 40 at r2](https://reviewable.io:443/reviews/igorleonenko/matches/5#-KT8w9HGb8BiodCWh1I7:-KT9GmdtzwF7u9dgjIgy:brloi2m) ([raw file](https://github.com/igorleonenko/matches/blob/15b185f040dd953810b7bdbcfcc6a3568086ff65/app/controllers/teams_controller.rb#L40)):_
<details><summary>

<i>Previously, mkarbowiak (Mateusz Karbowiak) wrote…</i></summary>



> You're using `user_ids` so shouldn't this variable be named `users`? Or is this just single id? 
> </details>
> Yes, in here its comes like single id

---

_[app/controllers/teams_controller.rb, line 53 at r2](https://reviewable.io:443/reviews/igorleonenko/matches/5#-KT9-4n-JRFU98X3ycD_:-KT9GtCi6VaONGA79Pfe:b-bkehg2) ([raw file](https://github.com/igorleonenko/matches/blob/15b185f040dd953810b7bdbcfcc6a3568086ff65/app/controllers/teams_controller.rb#L53)):_
<details><summary>

<i>Previously, mkarbowiak (Mateusz Karbowiak) wrote…</i></summary>



> And what if we want to update the name of the team? Or are we not allowed to?
> </details>
> No, I not allow to change team name

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/igorleonenko/matches/5)_

**Comment by IgorLeonenko** [comment]
_[app/models/tournament.rb, line 45 at r2](https://reviewable.io:443/reviews/igorleonenko/matches/5#-KT9000vVKUl1ctSk7xA:-KT9Me1uBrQTgW-muaZH:b-ce2v2x) ([raw file](https://github.com/igorleonenko/matches/blob/15b185f040dd953810b7bdbcfcc6a3568086ff65/app/models/tournament.rb#L45)):_

> resign
> raise error if try to add more players than @tournament.players_in_team. But anyway if it's better to do via validations, I must change my code

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/igorleonenko/matches/5)_

**Comment by IgorLeonenko** [comment]
_[app/models/tournament.rb, line 45 at r2](https://reviewable.io:443/reviews/igorleonenko/matches/5#-KT9000vVKUl1ctSk7xA:-KT9NfDeONkOiRHXmS6F:bh3t2wj) ([raw file](https://github.com/igorleonenko/matches/blob/15b185f040dd953810b7bdbcfcc6a3568086ff65/app/models/tournament.rb#L45)):_
<details><summary>

<i>Previously, IgorLeonenko (Igor) wrote…</i></summary>



> > resign
> > raise error if try to add more players than @tournament.players_in_team. But anyway if it's better to do via validations, I must change my code
> > </details>
> > And of course in this case I must change code in controller

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/igorleonenko/matches/5)_
